src/edu/csus/ecs/pc2/clics/API202306/CLICSContestState.java

```diff
@@ -0,0 +1,149 @@
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import edu.csus.ecs.pc2.core.Utilities;
+import edu.csus.ecs.pc2.core.model.ContestInformation;
+import edu.csus.ecs.pc2.core.model.ContestTime;
+import edu.csus.ecs.pc2.core.model.FinalizeData;
+import edu.csus.ecs.pc2.core.model.IInternalContest;
+import edu.csus.ecs.pc2.core.util.IJSONTool;
+import edu.csus.ecs.pc2.services.core.JSONUtilities;
+
+/**
+ * CLICS Contest State
+ *
+ * @author John Buck
+ *
+ */
+// 2023-06 Spec specifically says to return a time or null, so always include all properties
+// But it also says the fields are optional, so, I assume, if it's null, that means the CCS
+// supports that property, which PC2 does.  If it did not support the property, then it would
+// not be included in the JSON.  (At least that is my interpretation -- JohnB)
+@JsonInclude(JsonInclude.Include.ALWAYS)
+public class CLICSContestState {
+
+    @JsonProperty
+    private String started;
+
+    @JsonProperty
+    private String frozen;
+
+    @JsonProperty
+    private String ended;
+
+    @JsonProperty
+    private String thawed;
+
+    @JsonProperty
+    private String finalized;
+
+    @JsonProperty
+    private String end_of_updates;
+
+    /*
+     * These are non-standard PC2 specific properties to support pausing the contest
+     * 'paused' is the CONTEST elapsed time at which the contest was paused.  If set, it means the contest
+     * is currently paused. (HH:MM:SS.uuu)
+     *
+     * 'resumed' is the real (wall) time the contest was last resumed (or started, including initial contest start).
+     * You can tell if the contest had been paused at some point if resumed != started (ISO time)
+     */
+    @JsonProperty
+    private String paused;
+
+    @JsonProperty
+    private String resumed;
+
+    /*
+     * Non-standard property supplying a client with the current contest time (HH:MM:SS.uuu)
+     */
+    @JsonProperty
+    private String contest_time;
+
+    /**
+     * Fill in properties for contest state as per 2023-06 spec
+     *
+     * @param model place to get contest times from
+     * @param ci optional contest information.  If null, use the model's info.  This constructor
+     *      is also used by ContestInformation objects that may come from an event listener, so
+     *      we make it variable as it may not be from the current model.
+     */
+    public CLICSContestState(IInternalContest model, ContestInformation ci) {
+
+        if (model.getContestTime().isContestStarted()) {
+            if(ci == null) {
+                ci = model.getContestInformation();
+            }
+            ContestTime ct = model.getContestTime();
+
+            started = Utilities.getIso8601formatterWithMS().format(ct.getContestStartTime().getTime());
+            if (model.getContestTime().isPastEndOfContest()) {
+                Calendar endedDate = IJSONTool.calculateElapsedWalltime(model, ct.getContestLengthMS());
+                if (endedDate != null) {
```

Comment:
This is an admittedly _bizzare_ edge case, but:  if a contest ends in the first minute, I believe `calculateElapsedWalltime()` will return `null`, which means `ended` will be unset (i.e., will be null) when it shouldn't be (it should be a time which is less than 60 seconds after the start...)

Not worth worrying about?

Reply:
Can the contest end in the first minute?  That would sound like a configuration error if start time == end time.  Of course, we can STOP the contest in the first minute, but that's different than ending.

However, I did notice that I could use "ct" instead of another call to model.getContestTime() in that block. :-)

Reply:
Well, it would _most likely_ be a configuration error, yes.  Like I said, _bizarre_ edge case.  Not worth worrying about...